### ISSUE TITLE
ps1 B

### ISSUE THREAD

**Issue by Mmorami** [John-L-Jones-IV/6.0002#1]
Hi @John-L-Jones-IV,

Im doing course 6.0002 and i've found your solution to compare mine with.
I liked your code and verbal answers, it shows you are at a high level.

I had some hard time with ps1B, and when i finally solved it (partially) i noticed that it assumes the cargo limit can always be reached. I then complicated my solution and actually returned some extra info that was'nt supposed to be returned by the function, but in the end it solved this problem.

Your algorithm does the same - it assumes that the list of egg weights will reach the limit exactly with at least 1 combination. I've found a test that our code fail to solve correctly if the limit can't be reached.
The following list cannot reach 30 with any combination:
```    
    egg_weights = (7, 11, 13) 
    n = 30
    print("Egg weights = (7, 11, 13)")
    print("n = 30")
    print("Expected ouput: 3 (2 * 11 + 1 * 7 = 29)")
    print("Actual output:", dp_make_weight(egg_weights, n))
    print()
```

You are more than welcome to see my solution (though it is definitely not as elegant as yours) at [my repository](https://github.com/Mmorami/MIT_Intro_to_Computational_Thinking_and_Data_Science_6.0002) under folder ps1, and the execution of my test on your code at tries.py

Cheers

**Comment by John-L-Jones-IV** [comment]
@Mmorami,

In the docstring provided for the dp_make_weight() function, we are told to assume an infinite supply of eggs for each weight and there is always a egg of value 1. I think that addresses our concern. The target weight of type int will always be completely satisfied if there are egg values of 1 around.

Note: this is a greedy solution and does not always offer the optimal solution which your test:

    egg_weights = (7, 11, 13)
    n = 30
    print("Egg weights = (7, 11, 13)")
    print("n = 30")
    print("Expected ouput: 3 (2 * 11 + 1 * 7 = 29)")
    print("Actual output:", dp_make_weight(egg_weights, n))

expects.

I think the course intends to bring awareness to the complexity of optimization problems without a computationally expensive brute force algorithm. Some quick googling has showed me the knapsack problem is NP-complete, which means some guys smarter than me have determined "there is no known algorithm both correct and fast (polynomial-time) in all cases".

[Knapsack Problem on Wikipedia](https://en.wikipedia.org/wiki/Knapsack_problem)

Anyways, I am initiating a pull request for your tries.py code with a small modification so it won't break when egg value 1 is not around.

Cheers,
J

**Comment by Mmorami** [comment]
You are right about the assumption of always having 1kg egg.
Indeed I skipped the docstring as it usually repeats the assignment pdf file.

I'm not sure what you meant in the note. I implemented a brute force that consider every possibility and chooses the optimal one (least amount of eggs). Yours is not quite a greedy algorithm either. It functions different than mine, but still cosider the least amount of steps, even if taking the heaviest egg is not included.
